randovania/game_connection/executor/msr_executor.py

```diff
@@ -0,0 +1,335 @@
+    PACKET_COLLECTED_INDICES = b"6"
+    PACKET_RECEIVED_PICKUPS = b"7"
+    PACKET_GAME_STATE = b"8"
+    PACKET_MALFORMED = b"9"
+
+
+class ClientInterests(IntEnum):
+    LOGGING = b"1"
+    MULTIWORLD = b"2"
+
+
+# FIXME: This is a copy of ODR's implementation just that the first param is a path instead of a name
+# for a file within ODR's template folder
+def replace_lua_template(file: Path, replacement: dict[str, Any], wrap_strings: bool = False) -> str:
+    from open_samus_returns_rando.misc_patches.lua_util import lua_convert  # type: ignore
+
+    code = file.read_text()
+    for key, content in replacement.items():
+        # Replace `TEMPLATE("key")`-style replacements
+        code = code.replace(f'TEMPLATE("{key}")', lua_convert(content, wrap_strings))
+        # Replace `T__key__T`-style replacements
+        code = code.replace(f"T__{key}__T", lua_convert(content, wrap_strings))
+
+    unknown_templates = re.findall(r'TEMPLATE\("([^"]+)"\)', code)
+    unknown_templates.extend(re.findall(r"T__(\w+)__T", code))
+
+    if unknown_templates:
+        raise ValueError("The following templates were left unfulfilled: " + str(unknown_templates))
+
+    return code
+
+
+# This is stupid but MSRExecutor defines a "connect" method which makes a lot of trouble if it would
+# inherit QObject
+class MSRExecutorToConnectorSignals(QObject):
+    new_inventory = Signal(str)
+    new_collected_locations = Signal(bytes)
+    new_received_pickups = Signal(str)
+    new_player_location = Signal(str)
+    connection_lost = Signal()
+
+
+def get_bootstrapper_for(game: GameDescription) -> list[str]:
+    all_code = []
+    bootstrap_path = game.game.data_path.joinpath("assets", "lua")
+    replacements = {
+        "num_pickup_nodes": game.region_list.num_pickup_nodes,
+        "inventory": "{{{}}}".format(
+            ",".join(repr(r.extra["item_id"]) for r in game.resource_database.item if "item_id" in r.extra)
+        ),
+    }
+
+    for i in range(4):
+        bootstrap_part = bootstrap_path.joinpath(f"bootstrap_part_{i}.lua")
+        all_code.append(replace_lua_template(bootstrap_part, replacements))
+
+    locations_lua = bootstrap_path.joinpath("bootstrap_locations.lua")
+    for world in game.region_list.regions:
+        entries = []
+
+        for node in world.all_nodes:
+            if isinstance(node, PickupNode):
+                if "actor_name" in node.extra:
+                    key = node.extra["actor_name"]
+                else:
+                    key = node.extra["spawngroup"]
+                entries.append(f"{key}={node.pickup_index.index + 1}")
+
+        if not entries:
+            continue
+
+        replacements["pairs"] = "{}".format(",".join(entries))
+        replacements["location"] = "{}".format(repr(world.extra["scenario_id"] + "_"))
+        code = replace_lua_template(locations_lua, replacements)
+        all_code.append(code)
+
+    return all_code
+
+
+class MSRExecutor:
+    _port = 42069
```

Comment:
Should we use a more unique port such as the release date (17915)

Reply:
Don't want to go through the hassle of what needs to be done for that.
Maybe whenever I have to update the remote connector again.